File: decisionTree.py
class Node:
    def __init__(self, feature=None, threshold=None, left=None, right=None, label=None):
        self.feature = feature
        self.threshold = threshold
        self.left = left
        self.right = right
        self.label = label  # only for leaves

    def is_leaf(self):
        return self.label is not None

def impurity(data):
    total = len(data)
    if total == 0:
        return 0
    counts = {}
    for row in data:
        label = row[-1]
        counts[label] = counts.get(label, 0) + 1
    gini = 1
    for c in counts.values():
        p = c / total
        gini -= p ** 2
    return gini

def weighted_impurity(left, right):
    n_left, n_right = len(left), len(right)
    n_total = n_left + n_right

    return (n_left/n_total) * impurity(left) + (n_right/n_total) * impurity(right)

def partition(data, j, t):
    left = [row for row in data if row[j] <= t]
    right = [row for row in data if row[j] > t]
    return left, right


def find_best_split(data, features):
    best_impurity = float('inf')
    best_j, best_t = None, None

    for j in features:
        values = sorted(set(row[j] for row in data)) # converts the set{} to list[]
        thresholds = [(values[i] + values[i+1]) / 2 for i in range(len(values)-1)]

        for t in thresholds:
            print(f"Finding best split among them: feat: {j}, thres: {t}:")

            left, right = partition(data, j, t)
            print(f"Partition: Left: {left}, Right: {right}")

            if not left or not right:
                continue
            w = weighted_impurity(left, right)
            print(f"Weighted impurity of child node: {w}\n")
            if w < best_impurity:
                best_impurity = w
                best_j, best_t = j, t

    return best_j, best_t, best_impurity


def majority_class(data):
    counts = {}
    for row in data:
        label = row[-1]
        counts[label] = counts.get(label, 0) + 1

    return max(counts, key=counts.get) # returns the most common class


def build_tree(data, features, depth=0, max_depth=5, min_samples=2):
    if depth >= max_depth or len(data) < min_samples or impurity(data) == 0:
        return Node(label=majority_class(data))

    j, t, best_impurity = find_best_split(data, features)
    print(f"\nbest feature: {j}, best_threshold: {t}, best_impurity: {best_impurity}")

    if j is None or impurity(data) - best_impurity <= 0:
        print("No better impurity")
        return Node(label=majority_class(data))

    left_data, right_data = partition(data, j, t)
    print(f"left_data: {left_data} \n Right_data: {right_data}\n")

    left_subtree = build_tree(left_data, features, depth+1, max_depth, min_samples)
    right_subtree = build_tree(right_data, features, depth+1, max_depth, min_samples)

    return Node(feature=j, threshold=t, left=left_subtree, right=right_subtree)


def predict(x, node):
    if node.is_leaf():
        return node.label
    if x[node.feature] <= node.threshold:
        return predict(x, node.left)
    else:
        return predict(x, node.right)

# usage
# feature_0 = Hours studied, feature_1 = Slept well (0/1), label = Pass (0/1)
# feature_0 = Income (k$), feature_1 = Credit score (hundreds, e.g. 6=600s), 
# feature_2 = Has existing debt (0/1), label = Loan approved (0/1)

train = [
    [25, 5, 1, 0],
    [30, 6, 1, 0],
    [45, 6, 0, 1],
    [20, 4, 1, 0],
    [60, 7, 0, 1],
    [35, 5, 1, 0],
    [50, 7, 1, 1],
    [22, 5, 0, 0],
    [70, 8, 0, 1],
    [40, 6, 1, 0],
    [55, 7, 0, 1],
    [28, 4, 1, 0],
    [65, 8, 1, 1],
    [33, 5, 0, 0],
    [48, 6, 0, 1],
    [38, 5, 1, 0],
    [58, 7, 1, 1],
    [26, 4, 0, 0],
    [42, 6, 1, 1],
    [31, 5, 1, 0],
]

test = [
    [27, 5, 1, 0],
    [52, 7, 0, 1],
    [44, 6, 1, 1],
    [24, 4, 0, 0],
    [63, 8, 0, 1],
    [36, 5, 1, 0],
]

# usage
num_features = len(train[0]) - 1   # exclude the label column
features = list(range(num_features))
tree = build_tree(train, features=features)

correct = 0
for row in test:
    x = row[:-1]
    true_label = row[-1]
    pred = predict(x, tree)
    correct += (pred == true_label)
    print(f"x={x}  true={true_label}  predicted={pred}")

print(f"Accuracy: {correct}/{len(test)}")

File: decisionTree.cpp
#include <iostream>
#include <vector>
using namespace std;

class Sample{
public:
    int age;
    int salary;
    int label;

    Sample(int age, int salary, int l){
        this->age = age;
        this->salary = salary;
        this->label = l;
    }
};


class Node{
public:
    int featureIndex; // which feature to split on
    double threshold; // Split value

    int prediction; // used only for leaf nodes
    bool isLeaf;

    Node* left;
    Node* right;

    Node(){
        featureIndex = -1;
        threshold = 0;

        isLeaf = false;
        prediction = -1;

        left = NULL;
        right = NULL;
    }
};

class Split{
public:
    int featureIndex;
    int threshold;
    double score;

    Split(){
        featureIndex = -1;
        threshold = 0;
        score = 999999; // start with bad score
    }
};

class DecisionTree{
private:
    Node* root;
    
    // Recursive prediction
    int predict(Node* node, int age, int salary){
        if(node->isLeaf){
            return node->prediction;
        }
        if(node->featureIndex == 0){
            if(age <= node->threshold) return predict(node->left, age, salary);
            else return predict(node->right, age, salary);
        }
        return -1;
    }

    double gini(vector<Sample>& data){
        int countNo = 0;
        int countYes = 0;

        for(const Sample& sample: data){
            if(sample.label == 0){
                countNo++;
            }
            else{
                countYes++;
            }
        }

        double total = data.size();

        double pNo = countNo / total;
        double pYes = countYes / total;

        return 1.0 - (pNo*pNo + pYes*pYes);
    }

    double scoreSplit(vector<Sample>& left, vector<Sample>& right){
        int total = left.size() + right.size();

        double leftWeight = (double) left.size() / total;
        double rightWeight = (double) right.size() / total;

        return leftWeight*gini(left) + rightWeight*gini(right);
    }

    int majorityClass(vector<Sample>& data){
        int countNo = 0;
        int countYes = 0;

        for(const Sample& sample: data){
            if(sample.label == 0){
                countNo++;
            }
            else countYes++;
        }
        if (countYes > countNo) return 1;
        
        return 0;
    }


    Split findBestSplit(vector<Sample>& data){
        Split best;

        // Try every feature
        for(int feat = 0; feat < 2; feat++){
            for(const Sample& sample: data){
                int threshold;

                if(feat == 0) threshold = sample.age;
                else threshold = sample.salary;

                vector<Sample> left;
                vector<Sample> right;

                // Divide the dataset
                for(const Sample& s: data){
                    int value;

                    if(feat == 0) value = s.age;
                    else value = s.salary;
                    
                    if(value <= threshold) left.push_back(s);
                    else right.push_back(s);
                }
                if(left.empty() || right.empty()) continue;

                double score = scoreSplit(left, right);

                if(score < best.score){
                    best.featureIndex = feat;
                    best.threshold = threshold;
                    best.score = score;
                }
            }
        }
        return best;
    }

    Node* buildTree(vector<Sample>& data){
        Node* node = new Node();

        bool same = true;

        for(int i=1; i<data.size(); i++){
            if(data[i].label != data[0].label){
                same = false;
                break;
            }
        }
        // base case
        if(same){
            node->isLeaf = true;
            node->prediction = data[0].label;
            return node;
        }

        Split best = findBestSplit(data);

        if(best.featureIndex == -1){
            node->isLeaf = true;
            node->prediction = majorityClass(data);

            return node;
        }

        node->threshold = best.threshold;
        node->featureIndex = best.featureIndex;

        vector<Sample> leftData;
        vector<Sample> rightData;

        for(const Sample& s: data){
            int val;
            if(best.featureIndex == 0) val = s.age;
            else val = s.salary;

            if(val <= best.threshold){
                leftData.push_back(s);
            }
            else{
                rightData.push_back(s);
            }
        }

        node->left = buildTree(leftData);
        node->right = buildTree(rightData);

        return node;

    }

    int predict(Node* node, int age, int salary){
        if(node->isLeaf){
            return node->prediction;
        }

        int value;

        if(node->featureIndex == 0){
            value = age;
        }
        else value = salary;

        if(value <= node->threshold){
            return predict(node->left, age, salary);
        }
        else{
            return predict(node->right, age, salary);
        }
    }


public:
    DecisionTree(){
        root = NULL;
    }
    void train(vector<Sample>& data){
        root = buildTree(data);
    }
    int predict(int age, int salary){
        return predict(root, age, salary);
    }

};




int main()
{
    vector<Sample> data;


    data.push_back(Sample(22,25000,0));
    data.push_back(Sample(25,30000,0));
    data.push_back(Sample(28,40000,0));

    data.push_back(Sample(35,60000,1));
    data.push_back(Sample(40,70000,1));
    data.push_back(Sample(45,80000,1));



    DecisionTree tree;


    tree.train(data);



    int result = tree.predict(27,35000);



    if(result == 1)
        cout<<"Buy Laptop";
    else
        cout<<"Don't Buy Laptop";


    return 0;
}